test/e2e/apps/util.go

```diff
@@ -0,0 +1,109 @@
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package apps
+
+import (
+	"context"
+	"fmt"
+	"maps"
+	"strconv"
+	"sync"
+
+	"github.com/onsi/ginkgo/v2"
+	v1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/util/wait"
+	"k8s.io/apimachinery/pkg/watch"
+	clientset "k8s.io/client-go/kubernetes"
+	"k8s.io/client-go/tools/cache"
+	watchtools "k8s.io/client-go/tools/watch"
+	"k8s.io/klog/v2"
+	"k8s.io/kubernetes/test/e2e/framework"
+)
+
+type IndexedPodAnnotationTracker struct {
+	sync.Mutex
+	ownerName            string
+	ownerNs              string
+	labelSelector        string
+	podIndexAnnotation   string
+	podTrackedAnnotation string
+	trackedAnnotations   map[int][]string
+}
+
+func NewIndexedPodAnnotationTracker(ownerName, ownerNs, labelSelector, podIndexAnnotation, podTrackedAnnotation string) *IndexedPodAnnotationTracker {
+	return &IndexedPodAnnotationTracker{
+		ownerName:            ownerName,
+		ownerNs:              ownerNs,
+		labelSelector:        labelSelector,
+		podIndexAnnotation:   podIndexAnnotation,
+		podTrackedAnnotation: podTrackedAnnotation,
+		trackedAnnotations:   make(map[int][]string),
+	}
+}
+
+func (t *IndexedPodAnnotationTracker) Start(ctx context.Context, c clientset.Interface) context.CancelFunc {
+	ownerKey := klog.KRef(t.ownerNs, t.ownerName)
+	podClient := c.CoreV1().Pods(t.ownerNs)
+	podsList, err := podClient.List(ctx, metav1.ListOptions{LabelSelector: t.labelSelector})
+	framework.ExpectNoError(err, "failed to list Pods")
+
+	trackerCtx, trackerCancel := context.WithCancel(ctx)
+
+	go func() {
+		defer ginkgo.GinkgoRecover()
+
+		w := &cache.ListWatch{
+			WatchFunc: func(options metav1.ListOptions) (watch.Interface, error) {
+				options.LabelSelector = t.labelSelector
+				return podClient.Watch(ctx, options)
+			},
+		}
+
+		ginkgo.By(fmt.Sprintf("Start the Pod watch for owner: %s", ownerKey))
+		_, err = watchtools.Until(trackerCtx, podsList.ResourceVersion, w, func(event watch.Event) (bool, error) {
+			if event.Type == watch.Added {
+				if pod, ok := event.Object.(*v1.Pod); ok {
+					framework.Logf("Observed event for Pod %q with index=%v, annotation value=%v",
+						klog.KObj(pod), pod.Annotations[t.podIndexAnnotation], pod.Annotations[t.podTrackedAnnotation])
+					podIndex, err := strconv.Atoi(pod.Annotations[t.podIndexAnnotation])
+					if err != nil {
+						return true, fmt.Errorf("failed to parse pod index for Pod %q: %v", klog.KObj(pod), err.Error())
+					}
+					t.Lock()
+					defer t.Unlock()
+					t.trackedAnnotations[podIndex] = append(t.trackedAnnotations[podIndex], pod.Annotations[t.podTrackedAnnotation])
```

Comment:
we could check that the current value is an increment of the old value

Reply:
We could, but I wanted the tracker to be potentially more reusable for other contexts. If the new value is not old value +1, then simply this assert will fail : https://github.com/kubernetes/kubernetes/pull/130390/files#diff-909ebfa7fcb2a94030da3371d35863a2d723fc93356334e35f68816f5e313379R699, at the end of the test.

Reply:
Right, it is not needed as we basically test that in the final assert and we can keep it reusable.